metricflow-semantics/metricflow_semantics/time/dateutil_adjuster.py

```diff
@@ -0,0 +1,114 @@
+            return relativedelta(days=count)
+        elif time_granularity is TimeGranularity.WEEK:
+            return relativedelta(weeks=count)
+        elif time_granularity is TimeGranularity.MONTH:
+            return relativedelta(months=count)
+        elif time_granularity is TimeGranularity.QUARTER:
+            return relativedelta(months=count * 3)
+        elif time_granularity is TimeGranularity.YEAR:
+            return relativedelta(years=count)
+        else:
+            assert_values_exhausted(time_granularity)
+
+    @override
+    def expand_time_constraint_to_fill_granularity(
+        self, time_constraint: TimeRangeConstraint, granularity: TimeGranularity
+    ) -> TimeRangeConstraint:
+        adjusted_start = self.adjust_to_start_of_period(granularity, time_constraint.start_time)
+        adjusted_end = self.adjust_to_end_of_period(granularity, time_constraint.end_time)
+
+        if adjusted_start < TimeRangeConstraint.ALL_TIME_BEGIN():
+            adjusted_start = TimeRangeConstraint.ALL_TIME_BEGIN()
+        if adjusted_end > TimeRangeConstraint.ALL_TIME_END():
+            adjusted_end = TimeRangeConstraint.ALL_TIME_END()
+
+        return TimeRangeConstraint(start_time=adjusted_start, end_time=adjusted_end)
+
+    @override
+    def adjust_to_start_of_period(
+        self, time_granularity: TimeGranularity, date_to_adjust: datetime.datetime
+    ) -> datetime.datetime:
+        if time_granularity is TimeGranularity.DAY:
+            return date_to_adjust
+        elif time_granularity is TimeGranularity.WEEK:
+            return date_to_adjust + relativedelta(weekday=dateutil.relativedelta.MO(-1))
+        elif time_granularity is TimeGranularity.MONTH:
+            return date_to_adjust + relativedelta(day=1)
+        elif time_granularity is TimeGranularity.QUARTER:
+            if date_to_adjust.month <= 3:
+                return date_to_adjust + relativedelta(month=1, day=1)
+            elif date_to_adjust.month <= 6:
+                return date_to_adjust + relativedelta(month=4, day=1)
+            elif date_to_adjust.month <= 9:
+                return date_to_adjust + relativedelta(month=7, day=1)
+            else:
+                return date_to_adjust + relativedelta(month=10, day=1)
+        elif time_granularity is TimeGranularity.YEAR:
+            return date_to_adjust + relativedelta(month=1, day=1)
+        else:
+            assert_values_exhausted(time_granularity)
+
+    @override
+    def adjust_to_end_of_period(
+        self, time_granularity: TimeGranularity, date_to_adjust: datetime.datetime
+    ) -> datetime.datetime:
+        if time_granularity is TimeGranularity.DAY:
+            return date_to_adjust
+        elif time_granularity is TimeGranularity.WEEK:
+            return date_to_adjust + relativedelta(weekday=dateutil.relativedelta.SU(1))
+        elif time_granularity is TimeGranularity.MONTH:
+            return date_to_adjust + relativedelta(day=31)
```

Comment:
Thanks for putting that docstring note on the class, I would've wondered about this.